clang/lib/CodeGen/Targets/RISCV.cpp

```diff
@@ -632,6 +633,29 @@ class RISCVTargetCodeGenInfo : public CommonCheriTargetCodeGenInfo {
 
   void setTargetAttributes(const Decl *D, llvm::GlobalValue *GV,
                            CodeGen::CodeGenModule &CGM) const override {
+    const auto *VD = dyn_cast_or_null<VarDecl>(D);
+    if (VD) {
+      if (VD->hasAttr<CHERIMMIODeviceAttr>()) {
+        auto *Attr = D->getAttr<CHERIMMIODeviceAttr>();
+        if (auto *GVar = llvm::dyn_cast<llvm::GlobalVariable>(GV)) {
+          llvm::cheri::CapabilityImportAttr CapImportAttr(
+              "mem", Attr->getDeviceName(), Attr->getEncodedPermissions(),
+              "export_");
+          GVar->addAttribute(CapImportAttr.getAttrName(), CapImportAttr.str());
+        }
+      }
+
+      if (VD->hasAttr<CHERISharedObjectAttr>()) {
+        auto *Attr = D->getAttr<CHERISharedObjectAttr>();
+        if (auto *GVar = llvm::dyn_cast<llvm::GlobalVariable>(GV)) {
```

Comment:
This check is shared between the two cases. Can you extract it so it's not duplicated? That might allow you to reduce the control flow nesting overall.

Reply:
https://github.com/CHERIoT-Platform/llvm-project/pull/147/commits/730e14ac9ac441112cb249c017943195245f0123#diff-2412b0407edd59fe07f0e2547e92c7e79fe352bb491447240b96d4272e62e035R638 lifts that check one level up, shared between the two cases.